openslides_backend/action/actions/assignment_candidate/mixins.py

```diff
@@ -0,0 +1,45 @@
+from typing import Any, Dict
+
+from ....permissions.permission_helper import has_perm
+from ....permissions.permissions import Permissions
+from ....shared.exceptions import PermissionDenied
+from ....shared.patterns import Collection, FullQualifiedId
+from ...action import Action
+
+
+class PermissionMixin(Action):
+    def check_permissions(self, instance: Dict[str, Any]) -> None:
+        if "user_id" in instance:
+            user_id = instance["user_id"]
+            assignment_id = instance["assignment_id"]
+        else:
+            assignment_candidate = self.datastore.get(
+                FullQualifiedId(Collection("assignment_candidate"), instance["id"]),
+                ["user_id", "assignment_id"],
+            )
+            user_id = assignment_candidate["user_id"]
+            assignment_id = assignment_candidate["assignment_id"]
+        assignment = self.datastore.get(
+            FullQualifiedId(Collection("assignment"), assignment_id),
+            ["meeting_id", "phase"],
+        )
+        meeting_id = assignment["meeting_id"]
+
+        missing_permissions = []
+        if self.user_id == user_id:
+            permission = Permissions.Assignment.CAN_NOMINATE_SELF
+            if not has_perm(self.datastore, self.user_id, permission, meeting_id):
+                missing_permissions.append(permission)
+        else:
+            permission = Permissions.Assignment.CAN_NOMINATE_OTHER
+            if not has_perm(self.datastore, self.user_id, permission, meeting_id):
+                missing_permissions.append(permission)
+        if assignment.get("phase") == "voting":
+            permission = Permissions.Assignment.CAN_MANAGE
+            if not has_perm(self.datastore, self.user_id, permission, meeting_id):
+                missing_permissions.append(permission)
```

Comment:
Mmh, the appending is not quite right here: Assume you have an assignment in phase "voting", you don't have `CAN_NOMINATE_SELF` and try to nominate yourself. Then the first if clause will trigger as well as the last one, leading to the error saying that `CAN_NOMINATE_SELF` and `CAN_MANAGE` are missing, when in reality only the latter is important. I think you can drop the array altogether, move the assignment phase check up (since this is the "more general" check) and just instantly throw the error when a permission is missing.

Reply:
Okay, changed this. I moved the phase part at the first part and updated the exception code.